pkg/controller/seed-controller-manager/addon/addon_controller.go

```diff
@@ -540,6 +543,31 @@ func (r *Reconciler) migrateHetznerCSIDriver(ctx context.Context, log *zap.Sugar
 	return nil
 }
 
+// Between v2.23 and v2.24, there was a change to the vSphere CSI driver immutable field volumeLifecycleModes
+// as a result, the CSDriver resource has to be redeployed
+// https://github.com/kubermatic/kubermatic/issues/12801
+func (r *Reconciler) migrateVsphereCSIDriver(ctx context.Context, log *zap.SugaredLogger, cluster *kubermaticv1.Cluster) error {
+	cl, err := r.KubeconfigProvider.GetClient(ctx, cluster)
+	if err != nil {
+		return fmt.Errorf("failed to get kube client: %w", err)
+	}
+
+	driver := &storagev1.CSIDriver{}
+	if err := cl.Get(ctx, types.NamespacedName{Name: "csi.vsphere.vmware.com"}, driver); apierrors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to get CSIDriver: %w", err)
+	}
+
+	if len(driver.Spec.VolumeLifecycleModes) > 1 || (len(driver.Spec.VolumeLifecycleModes) == 1 && driver.Spec.VolumeLifecycleModes[0] != storagev1.VolumeLifecyclePersistent) {
```

Comment:
nit: afaict in 2.24 KKP will try to install empty `CSIDriver`
https://github.com/kubermatic/kubermatic/blob/v2.24.0-beta.1/addons/csi/vsphere/vsphere-csi-driver.yaml#L31-L38

shouldn't it just check for larger than zero?
```suggestion
	if len(driver.Spec.VolumeLifecycleModes) > 0 {
```

Reply:
> `The CSIDriver "csi.vsphere.vmware.com" is invalid: spec.volumeLifecycleModes: Invalid value: []storage.VolumeLifecycleMode{"Persistent"}: field is immutable"`

As per the error message from #12801:

My understanding is that if the field is not set, it gets defaulted to `["Persistent"]`. We therefore don't want to run the migration routine on that scenario, because it's already the expected outcome.